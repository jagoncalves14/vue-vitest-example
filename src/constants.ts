export const BASE_IMAGE_URL = 'https://image.tmdb.org/t/p/w780'
export const API_URL = `https://api.themoviedb.org/3/discover/movie?api_key=${import.meta.env.VITE_APP_API_KEY}&language=en-US&sort_by=popularity.desc`
