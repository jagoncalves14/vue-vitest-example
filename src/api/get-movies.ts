import axios, { AxiosResponse } from 'axios'
import { TMovieData } from '@/types/api'
import { API_URL } from '@/constants'

export default async function getMoviesList(): Promise<TMovieData[]> {
  const response: AxiosResponse = await axios.get(API_URL)
  return response.data.results
}
