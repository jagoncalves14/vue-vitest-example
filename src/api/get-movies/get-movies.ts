import axios, { AxiosResponse } from 'axios'
import { TMovieData } from '@/types/api'
import { API_URL } from '@/constants'

export default async function getMovies(): Promise<TMovieData[]> {
  const response: AxiosResponse = await axios.get(
    `${API_URL}/discover/movie?api_key=${import.meta.env.VITE_APP_API_KEY}&language=en-US&sort_by=popularity.desc`
  )
  return response.data.results
}
