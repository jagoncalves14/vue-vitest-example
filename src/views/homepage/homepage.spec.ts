import Homepage from './homepage.vue'
import { mount, flushPromises } from '@vue/test-utils'
import getMovies from '@/api/get-movies'
import { getMoviesMock } from '@/api/__mocks__/get-movies'
import { BASE_IMAGE_URL } from '@/constants'

vi.mock('@/api/get-movies')

describe('Homepage', () => {
  describe.skip('Snapshots', () => {
    it('should mount correctly - with data', async () => {
      const wrapper = mount(Homepage)

      await flushPromises()

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('should mount correctly - with just one movie', async () => {
      // @ts-ignore
      getMovies.mockImplementationOnce(() => {
        return Promise.resolve([getMoviesMock.results[0]])
      })

      const wrapper = mount(Homepage)

      await flushPromises()

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('should mount correctly - without movies results', async () => {
      // @ts-ignore
      getMovies.mockImplementationOnce(() => {
        return Promise.resolve([])
      })

      const wrapper = mount(Homepage)

      await flushPromises()

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('Mounted', () => {
    it('should populate data - with success', async () => {
      const context = {
        data: [],
        updateNotificationsTimer: vi.fn(),
      }

      await Homepage.mounted?.call(context)

      expect(getMovies).toHaveBeenCalled()
      expect(context.data).toEqual(getMoviesMock.results)
    })

    it('should populate data - with error', () => {
      const context = {
        data: [],
        updateNotificationsTimer: vi.fn(),
      }

      // @ts-ignore
      getMovies.mockImplementationOnce(() => {
        return Promise.reject('Some error')
      })

      Homepage.mounted?.call(context)
      expect(getMovies).toHaveBeenCalled()
      expect(context.data).toEqual([])
    })
  })

  describe('Methods', () => {
    describe('setActiveMovie', () => {
      const context = {
        activeMovieId: 0,
      }

      Homepage.methods?.setActiveMovie.call(context, 123)

      expect(context.activeMovieId).toBe(123)
    })

    it('updateNotificationsTimer', async () => {
      vi.useFakeTimers()

      const context = {
        timer: 0,
      }

      Homepage.methods?.updateNotificationsTimer.call(context)

      vi.runAllTimers()

      expect(clearTimeout).toBeCalled()
      expect(setTimeout).toBeCalled()
      expect(context.timer).toBe(10000)

      vi.useRealTimers()
    })
  })

  describe('Watch', () => {
    describe('data', () => {
      it('should call setActiveMovieId when has results', () => {
        const context = {
          setActiveMovie: vi.fn(),
        }

        // @ts-ignore
        Homepage.watch?.data?.handler.call(context, getMoviesMock.results)

        expect(context.setActiveMovie).toBeCalledWith(getMoviesMock.results[0].id)
      })

      it('should NOT call setActiveMovieId when has no results', () => {
        const context = {
          setActiveMovie: vi.fn(),
        }

        // @ts-ignore
        Homepage.watch?.data?.handler.call(context, [])

        expect(context.setActiveMovie).not.toHaveBeenCalled()
      })
    })

    it('activeMovieId', () => {
      const context = {
        data: getMoviesMock.results,
        preview: {},
      }
      context.data[0].id = 123

      // @ts-ignore
      Homepage.watch?.activeMovieId?.handler.call(context, 123)

      expect(context.preview).toEqual({
        image: `${BASE_IMAGE_URL}${context.data[0].poster_path}`,
        title: context.data[0].title,
        overview: context.data[0].overview,
      })
    })
  })
})
