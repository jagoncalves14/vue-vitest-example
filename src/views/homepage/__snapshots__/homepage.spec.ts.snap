// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage Snapshots should mount correctly - with data 1`] = `
<div class="homepage-container">
  <div class="movie-list"><input class="movie-list__search" type="text" placeholder="Search for a movie">
    <ul>
      <li class="active">Morbius</li>
      <li class="">The Lost City</li>
      <li class="">Sonic the Hedgehog 2</li>
      <li class="">The Northman</li>
      <li class="">The Exorcism of God</li>
      <li class="">Uncharted</li>
      <li class="">The Batman</li>
      <li class="">Doctor Strange in the Multiverse of Madness</li>
      <li class="">Spider-Man: No Way Home</li>
      <li class="">The Bad Guys</li>
      <li class="">The Contractor</li>
      <li class="">Turning Red</li>
      <li class="">Moonfall</li>
      <li class="">The Takedown</li>
      <li class="">Virus:32</li>
      <li class="">Ambulance</li>
      <li class="">Doctor Strange</li>
      <li class="">The Outfit</li>
      <li class="">Fortress: Sniper's Eye</li>
      <li class="">Encanto</li>
    </ul>
  </div>
  <div class="movie-preview">
    <div class="movie-preview__image"><img src="https://image.tmdb.org/t/p/w780/6JjfSchsU6daXk2AKX8EEBjO3Fm.jpg"></div>
    <h3 class="movie-preview__title">Morbius</h3>
    <p class="movie-preview__title">Budget: $70,000,000.00</p>
  </div>
</div>
`;

exports[`Homepage Snapshots should mount correctly - with just one movie 1`] = `
<div class="homepage-container">
  <div class="movie-list"><input class="movie-list__search" type="text" placeholder="Search for a movie">
    <ul>
      <li class="active">Morbius</li>
    </ul>
  </div>
  <div class="movie-preview">
    <div class="movie-preview__image"><img src="https://image.tmdb.org/t/p/w780/6JjfSchsU6daXk2AKX8EEBjO3Fm.jpg"></div>
    <h3 class="movie-preview__title">Morbius</h3>
    <p class="movie-preview__title">Budget: $70,000,000.00</p>
  </div>
</div>
`;

exports[`Homepage Snapshots should mount correctly - without movies results 1`] = `
<div class="homepage-container">
  <div class="movie-list"><input class="movie-list__search" type="text" placeholder="Search for a movie">
    <!--v-if-->
  </div>
  <div class="movie-preview">
    <h3 class="movie-preview__loading">Loading...</h3>
  </div>
</div>
`;
